lib/TH/generic/THTensorMath.c

```diff
@@ -1057,64 +1057,135 @@ void THTensor_(reshape)(THTensor *r_, THTensor *t, THLongStorage *size)
 }
 
 /* I cut and pasted (slightly adapted) the quicksort code from
-   http://www.alienryderflex.com/quicksort/
-   This public-domain C implementation by Darel Rex Finley.
-   Thanks man :)
+   Sedgewick's 1978 "Implementing Quicksort Programs" article
+   http://www.csie.ntu.edu.tw/~b93076/p847-sedgewick.pdf
 
-    Updated Oct 16 2013: change choice of pivot to avoid worst-case being a pre-sorted input - Daniel and Julien
-    Updated Oct 24 2013: change pivot comparison to strict inequality to avoid worst-case on constant input, see Sedgewick, Algorithms in C, Addison Wesley, 1990, p. 120 - Julien
+   It is the state of the art existing implementation. The macros 
+   are here to make as close a match as possible to the pseudocode of
+   Program 2 p.851
+
+   Note that other partition schemes exist, and are typically presented
+   in textbook, but those are less efficient. See e.g.
+   http://cs.stackexchange.com/questions/11458/quicksort-partitioning-hoare-vs-lomuto
+
+   Julien, November 12th 2013
 */
-#define  MAX_LEVELS  300
+#define MAX_LEVELS  300
+#define M_SMALL 10 /* Limit for small subfiles */
```

Comment:
this has not been undefined below anywhere. you might have to add a #undef around line 1179